iceoryx_posh/doc/chunk_header.md

```diff
@@ -0,0 +1,323 @@
+
+# Open Issues
+
+- the design was done with the intention to have a custom header of arbitrary size, if the size is limited to e.g. 32 bytes, some things could be simplified
+- typed publisher/subscriber API proposal
+```
+// publisher
+auto pub = iox::popo::TypedPublisher<MyPayload, MyHeader>(serviceDescription);
+pub.loan()
+    .and_then([&](auto& sample) {
+        sample.getHeader()->customHeader<MyHeader>().data = 42;
+        sample->a = 42;
+        sample->b = 13;
+        sample.publish();
+    })
+    .or_else([](iox::popo::AllocationError) {
+        // Do something with error.
+    });
+
+// subscriber
+auto sub = iox::popo::TypedSubscriber<MyPayload, MyHeader>(serviceDescription);
+sub->take()
+    .and_then([](auto& sample){
+        std::cout << "Custom header data: " << sample.getHeader()->customHeader<MyHeader>().data << std::endl;
+        std::cout << "Payload data: " << static_cast<const MyPayload*>(sample->get()).data << std::endl;
+    });
+```
+    - the publisher/subscriber would have a default parameter for the custom header to be source compatible with our current API
+    - the drawback is that the user could use the wrong custom header. Maybe `Sample` also needs an additional template parameter
+    - additionally, a `ChunkHeaderHook` could be used on the publisher side
+```
+template <typename CustomHeader>
+class MyChunkHeaderHook : public ChunkHeaderHook
+{
+  public:
+    void deliveryHook(ChunkHeader& chunkHeader) override
+    {
+        chunkHeader.customHeader<CustomHeader>().timestamp = myTimeProvider::now();
+    }
+};
+
+auto customHeaderHook = MyChunkHeaderHook<MyHeader>();
+auto pub = iox::popo::TypedPublisher<MyPayload>(serviceDescription, customHeaderHook);
+```
+        - it has to be decided if it's a good idea to give the user easy write access to the `ChunkHeader` or just to the `CustomHeader`
+- untyped publisher/subscriber API proposal
+```
+// publisher option 1
+auto pub = iox::popo::UntypedPublisher<MyHeader>(serviceDescription);
+
+// publisher option 2
+auto customHeaderSize = sizeOf(MyHeader);
+auto pub = iox::popo::UntypedPublisher(serviceDescription, customHeaderSize);
+
+auto payloadSize = sizeof(MyPayload);
+auto payloadAlignment = alignof(MyPayload);
+pub.loan(payloadSize, payloadAlignment)
+    .and_then([&](auto& sample) {
+        sample.getHeader()->customHeader<MyHeader>().data = 42;
+        auto payload = new (sample.get()) MyPayload();
+        payload->data = 73;
+        sample.publish();
+    })
+    .or_else([](iox::popo::AllocationError) {
+        // Do something with error.
+    });
+
+// subscriber option 1
+auto pub = iox::popo::UntypedPublisher<MyHeader>(serviceDescription);
+
+// subscriber option 2
+auto customHeaderSize = sizeOf(MyHeader);
+auto pub = iox::popo::UntypedSubscriber(serviceDescription, customHeaderSize);
```

Comment:
Maybe avoiding a template here makes sense. And default for customHeaderSize would be zero. Then there is no change in the API if the custom header is not used

Reply:
I would still lean towards option 1, since it's more save and I can't think of a use case to change the header between samples. The small change would be to use `UntypedPublisher<>(serviceDescription)` if the custom header is not used